t/helper/test-gvfs-protocol.c

```diff
@@ -0,0 +1,1191 @@
+			goto done;
+		strbuf_trim_trailing_newline(&h);
+
+		if (!h.len)
+			goto done; /* a blank line ends the header */
+
+		string_list_append(&req_data->header_list,
+				   strbuf_detach(&h, NULL));
+	}
+
+	/*
+	 * We do not attempt to read the <message-body>, if it exists.
+	 * We let our caller read/chunk it in as appropriate.
+	 */
+done:
+	if (trace2_is_enabled())
+	{
+		struct string_list_item *item;
+		trace2_printf("%s: %s", TR2_CAT, req_data->start_line.buf);
+		for_each_string_list_item(item, &req_data->start_line_fields)
+			trace2_printf("%s: Field: %s", TR2_CAT, item->string);
+		trace2_printf("%s: [uri-base '%s'][gvfs '%s'][args '%s' '%s']",
+			      TR2_CAT,
+			      req_data->uri_base.buf,
+			      req_data->gvfs_api.buf,
+			      req_data->slash_args.buf,
+			      req_data->quest_args.buf);
+		for_each_string_list_item(item, &req_data->header_list)
+			trace2_printf("%s: Hdrs: %s", TR2_CAT, item->string);
+	}
+
+	return 0;
+}
+
+static int dispatch(struct my_request_data *req_data)
+{
+	const char *method = req_data->start_line_fields.items[0].string;
+
+	if (!strcmp(req_data->gvfs_api.buf, "gvfs/objects")) {
+
+		if (!strcmp(method, "GET"))
+			return do__gvfs_objects__get(req_data);
+		if (!strcmp(method, "POST"))
+			return do__gvfs_objects__post(req_data);
+	}
+
+	logerror("TODO unsuppored GVFS Protocol REST API ('%s','%s')",
+		 method, req_data->gvfs_api.buf);
+	return -1;
+}
+
+static int execute(void)
+{
+	struct my_request_data req_data = MY_REQUEST_DATA__INIT;
+	char *addr = getenv("REMOTE_ADDR"), *port = getenv("REMOTE_PORT");
+	int err = 0;
+
+	if (addr)
+		loginfo("Connection from %s:%s", addr, port);
+
+	set_keep_alive(0);
+
+	while (1) {
+		my_request_data__release(&req_data);
+
+		alarm(init_timeout ? init_timeout : timeout);
+		err = my_request_data__read(&req_data, 0);
+		alarm(0);
+
+		if (err == 0)
+			err = dispatch(&req_data);
+
+		if (err == -1)
+			break;
+
+		if (err == 1) { /* let EOF exit without error */
+			err = 0;
+			break;
+		}
+	}
+
+	close(0);
+	close(1);
+
+	return err;
+}
+
+static int addrcmp(const struct sockaddr_storage *s1,
+    const struct sockaddr_storage *s2)
+{
+	const struct sockaddr *sa1 = (const struct sockaddr*) s1;
+	const struct sockaddr *sa2 = (const struct sockaddr*) s2;
+
+	if (sa1->sa_family != sa2->sa_family)
+		return sa1->sa_family - sa2->sa_family;
+	if (sa1->sa_family == AF_INET)
+		return memcmp(&((struct sockaddr_in *)s1)->sin_addr,
+		    &((struct sockaddr_in *)s2)->sin_addr,
+		    sizeof(struct in_addr));
+#ifndef NO_IPV6
+	if (sa1->sa_family == AF_INET6)
+		return memcmp(&((struct sockaddr_in6 *)s1)->sin6_addr,
+		    &((struct sockaddr_in6 *)s2)->sin6_addr,
+		    sizeof(struct in6_addr));
+#endif
+	return 0;
+}
+
+static int max_connections = 32;
+
+static unsigned int live_children;
+
+static struct child {
+	struct child *next;
+	struct child_process cld;
+	struct sockaddr_storage address;
+} *firstborn;
```

Comment:
Technically, after the first child exited, `firstborn` won't point at the first-born anymore, right?